### PR TITLE
refactor(engine): split tip pick-up/drop into virtualizable dependency

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR closes RCORE-625 by splitting the tip-handling procedures from `protocol_engine.execution.PipettingHandler` into a new handler: `TipHandler`.

The `TipHandler` in an abstract interface with two concrete implementations:

- `HardwareTipHandler` - handle tips using the Hardware API
- `VirtualTipHandler` - handle tips without touching the Hardware API

This PR makes various changes in order to support splitting these procedures into their own dependency that gets swapped out where appropriate.

## Changelog

The biggest change involved splitting up the `PipettingHandler` to remove tip handling logic. Logic was moved to:

- `PickUpTipImplementation`
- `DropTipImplementation`
- `TipHandler` (new interface)

The new `TipHandler` simply runs tip handling routines **in place**. In other words, it's the job of the parent collaborator to call `MovementHandler.move_to_well` first.

This ends up dividing work up quite nicely and removes a layer of indirection when reading the `PickUpTip` and `DropTip` commands.

## Risk assessment

Medium, especially with changes in the LegacyCommandMapper

## Test Plan

Keep an eye out for bugs in the following areas:

- General protocol tip handling
- Drop-tip after run complete
- Drop-tip after run cancel

We need to test on both PAPI 2.14 and <= 2.13, due to changes in the LegacyCommandMapper.

In all the following tests

- All protocols should analyze successfully
- After the protocol completes or is cancelled, the tip should be dropped in the trash

| Test                    | API version | Procedure                                      | Status        |
| ----------------------- | ----------- | ---------------------------------------------- | ------------- |
| Legacy protocol success | 2.13        | Run protocol to completion                     | Needs testing |
| Legacy protocol cancel  | 2.13        | Cancel protocol after a tip has been picked up | Needs testing |
| Python protocol success | 2.14        | Run protocol to completion                     | Needs testing |
| Python protocol cancel  | 2.14        | Cancle protocol after a tip has been picked up | Needs testing |

```python
# tip_handler_test_213.py
from opentrons.types import Mount
from opentrons.protocol_api import ProtocolContext

metadata = {
    "protocolName": "Tip Handler Test (Legacy)",
    "apiLevel": "2.13",
}

PIPETTE = "p300_single"
PIPETTE_MOUNT = Mount.LEFT
TIP_RACK = "opentrons_96_tiprack_300ul"
TIP_RACK_SLOT = "5"

def run(protocol: ProtocolContext) -> None:
    tip_rack = protocol.load_labware(TIP_RACK, TIP_RACK_SLOT)
    pipette = protocol.load_instrument(PIPETTE, PIPETTE_MOUNT, [tip_rack])

    # pick up a tip from A1
    pipette.pick_up_tip()
    protocol.delay(seconds=5)

    # return the tip to A1
    pipette.return_tip()
    protocol.delay(seconds=5)

    # pick up the next tip
    pipette.pick_up_tip()
    protocol.delay(seconds=5)

    # drop the tip in the trash
    pipette.drop_tip()
    protocol.delay(seconds=5)

    # pick up the next tip
    pipette.pick_up_tip()
    protocol.delay(seconds=5)
```

```python
# tip_handler_test_214.py
from opentrons.types import Mount
from opentrons.protocol_api import ProtocolContext

metadata = {
    "protocolName": "Tip Handler Test",
    "apiLevel": "2.14",
}

PIPETTE = "p300_single"
PIPETTE_MOUNT = Mount.LEFT
TIP_RACK = "opentrons_96_tiprack_300ul"
TIP_RACK_SLOT = "5"

def run(protocol: ProtocolContext) -> None:
    tip_rack = protocol.load_labware(TIP_RACK, TIP_RACK_SLOT)
    pipette = protocol.load_instrument(PIPETTE, PIPETTE_MOUNT, [tip_rack])

    # pick up a tip from A1
    pipette.pick_up_tip()
    protocol.delay(seconds=5)

    # return the tip to A1
    pipette.return_tip()
    protocol.delay(seconds=5)

    # pick up the next tip
    pipette.pick_up_tip()
    protocol.delay(seconds=5)

    # drop the tip in the trash
    pipette.drop_tip()
    protocol.delay(seconds=5)

    # pick up the next tip
    pipette.pick_up_tip()
    protocol.delay(seconds=5)
```

### CODE REVIEW

**Comment by mcous** [comment]
Current dependencies on/for this PR:
* edge
  * **PR #12179** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/12179" target="_blank">Graphite</a> 
    * **PR #12186** <a href="https://app.graphite.dev/github/pr/Opentrons/opentrons/12186" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/Opentrons/opentrons/12186?utm_source=stack-comment).

File: api/src/opentrons/protocol_engine/state/geometry.py [hunk 1]
```diff
@@ -284,6 +279,8 @@ def get_nominal_tip_geometry(
 
         This geometry data is based solely on labware and pipette definitions and
         does not take calibrated tip lengths into account.
+
+        For calibrated data, use `PipettingHandler.get_calibrated_tip_geometry`.
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/geometry.py]:
```suggestion
```

File: api/src/opentrons/protocol_engine/execution/hardware_stopper.py [hunk 2]
```diff
@@ -38,11 +38,9 @@ def __init__(
             hardware_api=hardware_api,
             state_store=state_store,
         )
-
-        self._pipetting_handler = pipetting or PipettingHandler(
+        self._tip_handler = tip_handler or create_tip_handler(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/hardware_stopper.py]: Maybe the `HardwareStopper` should only get a `HardwareTipHandler`

File: api/src/opentrons/protocol_engine/execution/tip_handler.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+        Returns:
+            Tip geometry of the picked up tip.
+        """
+        ...
+
+    async def drop_tip(self, pipette_id: str, home_after: Optional[bool]) -> None:
+        """Drop the attached tip into the named location.
+
+        Pipette should be in place over the desitnation prior to calling this method.
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/tip_handler.py]:
```suggestion
        Pipette should be in place over the destination prior to calling this method.
```

File: api/src/opentrons/protocol_runner/legacy_context_plugin.py [hunk 4]
```diff
@@ -126,6 +131,25 @@ def _handle_equipment_loaded(self, load_info: LegacyLoadInfo) -> None:
         pe_action = pe_actions.UpdateCommandAction(command=pe_command)
         self._actions_to_dispatch.put(pe_action)
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/legacy_context_plugin.py]: Put after `if` to match engine `loadPipette` implementation

**Comment by TamarZanzouri** [comment]
Tested on an OT-2! worked as expected :rocket:

Test | API version | Procedure | Status 
-- | -- | -- | --
Legacy protocol success | 2.13 | Run protocol to completion  | OK
Legacy protocol cancel | 2.13 | Cancel protocol after a tip has been picked up | OK
Python protocol success | 2.14 | Run protocol to completion | OK
Python protocol cancel | 2.14 | Cancle protocol after a tip has been picked up | OK

**Review by sanni-t — APPROVED** [review]
LGTM from the in-person review.

File: api/src/opentrons/protocol_engine/state/pipettes.py [hunk 5]
```diff
@@ -411,7 +411,20 @@ def get_is_ready_to_aspirate(
 
     def get_attached_tip_labware_by_id(self) -> Dict[str, str]:
         """Get the tiprack ids of attached tip by pipette ids."""
-        return self._state.attached_tip_labware_by_id
+        return dict(self._state.attached_tip_labware_by_id)
+
+    def validate_tip_state(self, pipette_id: str, expected_has_tip: bool) -> None:
```
Comment (TamarZanzouri) [inline api/src/opentrons/protocol_engine/state/pipettes.py:416]: Nice!

**Review by TamarZanzouri — APPROVED** [review]
Nice work! LGTM and tested on real hardware!

**Pull request merged**